jlm/llvm/opt/alias-analyses/PointerObjectSet.cpp

```diff
@@ -209,8 +248,155 @@ SupersetOfAllPointeesConstraint::Apply(PointerObjectSet & set)
+  for (auto & argument : lambdaNode.fctarguments())
+  {
+    if (!is<PointerType>(argument.type()))
+      continue;
+
+    const auto argumentPO = set.GetRegisterPointerObject(argument);
+    set.GetPointerObject(argumentPO).MarkAsPointsToExternal();
+  }
+
+  // All results of pointer type need to be flagged as HasEscaped
+  for (auto & result : lambdaNode.fctresults())
+  {
+    if (!is<PointerType>(result.type()))
+      continue;
+
+    // Mark the register as escaped, which will propagate the escaped flag to all pointees
+    const auto resultPO = set.GetRegisterPointerObject(*result.origin());
+    set.GetPointerObject(resultPO).MarkAsEscaped();
+  }
+
+  return true;
+}
+
+// Connects function calls to every possible target function
+bool
+FunctionCallConstraint::Apply(PointerObjectSet & set)
+{
+  bool modified = false;
+
+  // When the function pointer can point to anything visible outside of the module
+  const auto handleCallingExternalFunction = [&]
+  {
+    // Mark all the call's inputs as escaped, and all the outputs as pointing to external
+    for (size_t n = 0; n < CallNode_.NumArguments(); n++)
+    {
+      // TODO: Remove n+1 when #295 is resolved
+      const auto & inputRegister = *CallNode_.Argument(n + 1)->origin();
+      if (!is<PointerType>(inputRegister.type()))
+        continue;
+
+      const auto inputRegisterPO = set.GetRegisterPointerObject(inputRegister);
+      modified |= set.GetPointerObject(inputRegisterPO).MarkAsEscaped();
+    }
+
+    for (size_t n = 0; n < CallNode_.NumResults(); n++)
+    {
+      const auto & outputRegister = *CallNode_.Result(n);
+      if (!is<PointerType>(outputRegister.type()))
+        continue;
+
+      const auto outputRegisterPO = set.GetRegisterPointerObject(outputRegister);
+      modified |= set.GetPointerObject(outputRegisterPO).MarkAsPointsToExternal();
+    }
+  };
+
+  // When the function pointer can point to a function imported from outside the module
+  const auto handleCallingImportedFunction = [&](const PointerObject::Index imported)
+  {
+    // FIXME: Add special handling of common library functions
+    // Otherwise we don't know anything about the function
+    handleCallingExternalFunction();
+  };
+
+  // When the function pointer can point to a lambda node in our module
+  const auto handleCallingLambdaFunction = [&](const PointerObject::Index lambda)
+  {
+    auto & lambdaNode = set.GetLambdaNodeFromFunctionMemoryObject(lambda);
+
+    // If the number of parameters or number of results doesn't line up,
+    // assume this is not the function we are calling.
+    // Note that the number of arguments and results include 3 state edges: memory, loop and IO.
+    // Varargs are properly handled, since they get merged by a valist_op node before the CallNode.
+    if (lambdaNode.nfctarguments() != CallNode_.NumArguments()
+        || lambdaNode.nfctresults() != CallNode_.NumResults())
+      return;
```

Comment:
I know we have an issue open for this right now: https://github.com/phate/jlm/issues/274

Ignoring this, do you see anywhere that the arguments/results of a direct(!) call do not line up with the lambda? I am wondering why this check is necessary?